packages/pools/src/concentrated/fetch-tick-data-provider.ts

```diff
@@ -0,0 +1,185 @@
+  protected _oneForZeroTicks: LiquidityDepth[] = [];
+
+  protected _zeroForOneBoundIndex = new Int(0);
+  protected _oneForZeroBoundIndex = new Int(0);
+
+  constructor(
+    protected readonly baseNodeUrl: string,
+    protected readonly poolId: string,
+    protected readonly nextTicksRampMultiplier = new Int(9),
+    protected readonly maxNumRequeriesPerDenom = 9,
+    protected readonly tickFetcher: (
+      poolId: string,
+      tokenInDenom: string,
+      boundTickIndex: string
+    ) => Promise<TickDepthsResponse> = async (
+      poolId,
+      tokenInDenom,
+      boundTickIndex
+    ) => {
+      if (fetch === undefined)
+        throw new Error("Fetch method must be available in the environment");
+
+      const res = await fetch(
+        `${this.baseNodeUrl}osmosis/concentratedliquidity/v1beta1/liquidity_net_in_direction?pool_id=${poolId}&token_in=${tokenInDenom}&use_cur_tick=true&bound_tick=${boundTickIndex}`
+      );
+      if (!res.ok) {
+        throw new Error("Failed to fetch liquidity net in direction");
+      }
+      return (await res.json()) as TickDepthsResponse;
+    }
+  ) {}
+
+  getTickDepthsTokenOutGivenIn(
+    pool: ConcentratedLiquidityPool,
+    token: { denom: string; amount: Int },
+    getMoreTicks = false
+  ): Promise<TickDepths> {
+    return this.requestInDirectionWithInitialTickBound(
+      pool,
+      pool.token0 === token.denom,
+      token.amount,
+      getMoreTicks
+    );
+  }
+
+  getTickDepthsTokenInGivenOut(
+    pool: ConcentratedLiquidityPool,
+    token: { denom: string; amount: Int },
+    getMoreTicks = false
+  ): Promise<TickDepths> {
+    return this.requestInDirectionWithInitialTickBound(
+      pool,
+      pool.token0 !== token.denom,
+      token.amount,
+      getMoreTicks
+    );
+  }
+
+  /** Fetch and return **additional** ticks in the desired direction. Maintains state to determine which ticks have already been fetched. */
+  protected async requestInDirectionWithInitialTickBound(
+    pool: ConcentratedLiquidityPool,
+    zeroForOne: boolean,
+    amount: Int,
+    getMoreTicks: boolean
+  ): Promise<TickDepths> {
+    if (pool.id !== this.poolId) throw new Error("Invalid pool");
+
+    // get the previous bound index used to fetch ticks
+    const prevBoundIndex = zeroForOne
+      ? this._zeroForOneBoundIndex
+      : this._oneForZeroBoundIndex;
+
+    const isMaxTicks = zeroForOne
+      ? prevBoundIndex.lte(minTick)
+      : prevBoundIndex.gte(maxTick);
+
+    const prevTicks = zeroForOne
+      ? this._zeroForOneTicks
+      : this._oneForZeroTicks;
+
+    // check if has fetched all ticks, if so return existing ticks
+    if (isMaxTicks) {
+      return {
+        allTicks: prevTicks,
+        isMaxTicks: true,
+      };
+    }
+
+    const tokenInDenom = zeroForOne ? pool.token0 : pool.token1;
+
+    const setLatestBoundTickIndex = (index: Int) => {
+      if (zeroForOne) {
+        this._zeroForOneBoundIndex = index;
```

Comment:
`requestinDirectionWithInitialTickBound` performs a few mutations. I suggest refactoring to use immutable data structures and avoid mutating `_zeroForOneBoundIndex`, `_oneForZeroTicks`, `_zeroForOneBoundIndex`, and `_oneForZeroBoundIndex`

Reply:
How so? These params are effectively a cache, as the provider may be requested tick depths within a previously fetched tick bound. That's how our mobx query store version of the provider works, and how it's intended to be used.

Reply:
I don't have an opinion on the mobx query store, but as a general principle, side effects that cause mutation of properties should be avoided because the instance is changed unexpectedly when these properties are changed in various methods. The state of the instance might become inconsistent depending on the order of function calls

It's hard for me to know if that would be an issue here, but one way to ensure that it isn't is to use a more 'functional programming' approach to make sure that we avoid side effects.

You could create a state class above, and then create a new instance of that state provider with your function calls in order to track state.

Reply:
I do prefer full separation of state and code with functional programming. 

But in this case, OOP is needed. Generally, the lifecycle of an objects data is considered to be the lifecycle of the instance. Here, without the object having to know; instances are created with CL pools and released with CL pools.

So, when the pools refresh, the pool is dropped from memory, with it's instances of providers.